psl-core/src/main/java/org/linqs/psl/application/inference/online/OnlineInference.java

```diff
@@ -0,0 +1,136 @@
+
+import org.linqs.psl.application.inference.InferenceApplication;
+import org.linqs.psl.application.inference.online.messages.OnlineMessage;
+import org.linqs.psl.application.inference.online.messages.actions.controls.Stop;
+import org.linqs.psl.application.inference.online.messages.responses.ActionStatus;
+import org.linqs.psl.database.Database;
+import org.linqs.psl.model.rule.Rule;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+
+public abstract class OnlineInference extends InferenceApplication {
+    private static final Logger log = LoggerFactory.getLogger(OnlineInference.class);
+
+    private OnlineServer server;
+    private boolean modelUpdates;
+    private boolean stopped;
+    private double objective;
+
+    protected OnlineInference(List<Rule> rules, Database database) {
+        super(rules, database);
+    }
+
+    protected OnlineInference(List<Rule> rules, Database database, boolean relaxHardConstraints) {
+        super(rules, database, relaxHardConstraints);
+    }
+
+    @Override
+    protected void initialize() {
+        stopped = false;
+        modelUpdates = true;
+        objective = 0.0;
+
+        startServer();
+
+        super.initialize();
+
+        termStore.ensureVariableCapacity(atomManager.getCachedRVACount() + atomManager.getCachedObsCount());
+    }
+
+    @Override
+    public void close() {
+        stopped = true;
+        closeServer();
+        super.close();
+    }
+
+    private void closeServer() {
+        if (server != null) {
+            server.close();
+            server = null;
+        }
+    }
+
+    private void startServer() {
+        server = new OnlineServer();
+        server.start();
+    }
+
+    protected void executeAction(OnlineMessage action) {
+        String response = null;
+
+        if (action.getClass() == Stop.class) {
+            response = doStop();
+        } else {
+            throw new IllegalArgumentException("Unsupported action: " + action.getClass().getName() + ".");
+        }
+
+        server.onActionExecution(action, new ActionStatus(action, true, response));
+    }
+
+    protected String doStop() {
+        stopped = true;
+        return "OnlinePSL inference stopped.";
+    }
+
+    /**
+     * Optimize if there were any modelUpdates since the last optimization.
+     */
+    private void optimize() {
+        if (!modelUpdates) {
+            return;
+        }
+
+        log.trace("Optimization Start");
+        objective = reasoner.optimize(termStore);
+        log.trace("Optimization End");
+
+        modelUpdates = false;
+    }
+
+    @Override
+    public double internalInference() {
+        // Initial round of inference.
+        optimize();
+
+        while (!stopped) {
+            OnlineMessage action = server.getAction();
+            if (action == null) {
+                continue;
+            }
+
+            try {
+                executeAction(action);
```

Comment:
Log (TRACE) here or in executeAction() about the action that is about to be executed.